### DISCUSSION TITLE
AsQueryable Paramterization?

### DISCUSSION THREAD

**Discussion by to11mtm** [linq2db/linq2db#4421] [Q&A]
With `AsQueryable` being greatly enhanced, we can now more easily use things like `InsertWithOutput` for client side data sets.

However, one drawback currently, is that When a set of data is passed in, the values are inlined. This can cause issues for extremely large statements however, as AFAIK we don't have any 'backtracking' logic like `BulkCopy` (mind you, I don't see how we could do so and have it semantically make sense.)

Is there an existing way via the API's to get the input values parameterized?

And if not, does anyone have input on a good place to start as far as trying to implement? (I don't have v5 pipeline quite understood yet, but am happy to give it a shot if team has ideas/suggestions 😇)

**Comment by sdanyliv** [comment]
Use `CreateTempTable` for now. Currently we are working on Version 6. It is big refactoring PR and almost everything was refactored and improved, so better do not touch `master` branch code.

**Reply by to11mtm** [reply]
We are now on V6 and I've made some new tests/observations...

It looks like now you can parameterize with `.SelectQuery` however it can get ugly;

By that, I mean; if you try to chain `.SelectQuery` and `.UnionAll`, you get SQL like:

```
SELECT
    @Id,
    @Data
UNION ALL
SELECT
    @Id_1,
    @Data_1
UNION ALL
SELECT
    @Id_2,
    @Data_2
UNION ALL
```

Which 'works' but is very hard on the eyes and may have other detriments compared to what happens when you do `.AsQueryable(this IEnumerable, IDataContext)` and the values are inlined, e.x.

```
SELECT
    [t1].[Id],
    [t1].[Data]
FROM
    (
        SELECT NULL [Id], NULL [Data] WHERE 1 = 0
        UNION ALL
        VALUES
            (0,'Data 0'), (1,'Data 1'), (2,'Data 2'), (3,'Data 3'), --etc
        ) [t1]
```

Being able to effectively parameterize parts, or even living with 'all' of the row being parameterized, lets us provide specialized code paths where, instead of having to repeatedly call `.InsertWithIdentity` we could call `.InsertWithOutputAsync` (while making sure our output clause returns the right columns to re-match to the other records to insert with next statement. The issue with using the `.AsQueryable` without parameters, is that when the rows are for something like an event sourced system where one column is a large binary value, it would be better to parameterize that binary value rather than bloat the SQL

Biggest win is it would allow users to more easily use `InsertWithOutput/Async` or `UpdateWithOutput/Async` for more complex scenarios without the weirdness/churn of SelectQuery pattern. (That said, I will note that `Sql.Parameter` makes `SelectQuery` a good bit less painful now at least.) 

I guess there are a few options that come to mind here, not sure which is better or worse? I put it together like this cause I remembered we had this discussion open and I didn't want to open a feature request/etc without at least getting feedback here (i.e. avoiding churn...)

#### Option 1: Somehow optimize the case of `.SelectQuery=>.UnionAll(.SelectQuery)` so that the output winds up more like the `.AsQueryable` case.

This would at least make the weirdness of actually doing the pattern more on the .NET side and at least SQL side is cleaner

#### Option 2: Add an overload `.AsQueryable(this IEnumerable, IDataContext, bool)` that parameterizes the whole shebang.

I don't like this one as much because it feels too blunt, or at the bare minimum the use of a `bool` is a smell to me.

#### Option 3: Add an overload along the lines of `.AsQueryable<T>(this IEnumerable<T>, IDataContext, Expression<Func<T,T>>)` where the expression is a constructor to the args you want to treat as parameter (or something like that)

Happy to consider a better API option here... but something that accomplishes the same (i.e. being able to specify which columns are parameters) would be the ideal state...

#### Option 4: There's a way to cleanly do this and I just don't realize it.

Again, in this case I'd be willing to update docs accordingly.

**Reply by to11mtm** [reply]
@sdanyliv Any thoughts on above? Sorry for ping but I'd like to hopefully work on or at least steer this based on direction now that 6.0 is out (but also don't want to overchurn opening issue since this discussion is open.)

**Reply by sdanyliv** [reply]
Please create a new issue with samples. We will review it and aim to address it in the next release.
